apps/src/music/progress/ProgressManager.ts

```diff
@@ -0,0 +1,129 @@
+// This file contains a generic ProgressManager which any lab can include,
+// if it wants to make progress without reloading the page.
+
+// Abstract class that validates a set of conditions. How
+// the validation works is up to the implementor.
+export abstract class Validator {
+  abstract shouldCheckConditions(): boolean;
+  abstract checkConditions(): void;
+  abstract conditionsMet(conditions: string[]): boolean;
+  abstract clear(): void;
+}
+
+// A validation inside the progression.
+interface Validation {
+  conditions: string[];
+  message: string;
+  next: boolean;
+}
+
+// The definition of the progression.
+interface Progression extends Object {
+  panels: {
+    text: string;
+    toolbox: {
+      [key: string]: string;
+    };
+    sounds: {
+      [key: string]: string;
+    };
+    validations: Validation[];
+  }[];
+}
+
+// The current progress state.
+interface ProgressState {
+  step: number;
+  satisfied: boolean;
+  message?: string | null;
+}
+
+export default class ProgressManager {
+  private progression: Progression;
+  private validator: Validator;
+  private onProgressChange: (state: ProgressState) => void;
+  private currentProgressState: ProgressState;
+
+  constructor(
+    progression: Progression,
+    validator: Validator,
+    onProgressChange: (state: ProgressState) => void
+  ) {
+    this.progression = progression;
+    this.validator = validator;
+    this.onProgressChange = onProgressChange;
+    this.currentProgressState = {
+      step: 0,
+      satisfied: false,
+      message: null
+    };
+  }
+
+  getProgression() {
+    return this.progression;
+  }
+
+  getCurrentState() {
+    return this.currentProgressState;
+  }
+
+  getCurrentToolbox() {
+    return this.progression.panels[this.currentProgressState.step].toolbox;
+  }
+
+  getCurrentSounds() {
+    return this.progression.panels[this.currentProgressState.step].sounds;
+  }
```

Comment:
Minor - would it be better to just expose the whole `panel` and let the consumer split it up into `toolbox`, `sounds`, etc? I'm thinking when this gets used for more labs, having multiple getters for lab-specific data here might start get unwieldy? Toolbox does seem fairly generic, but sounds is pretty musiclab specific right now

Reply:
I'm also going to rename `panels` to something like `steps`...